### PR TITLE
interpolate_to_different_mesh for hp::DoFHandler

### PR DESCRIPTION
**Pull request by starki0815**
I am offering this for incorporation into deal.ii. The patch (hopefully) makes interpolate_to_different_mesh work with hp::DoFHandler objects.

### CODE REVIEW

File: tests/numerics/interpolate_to_different_mesh_01.cc [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  VectorTools::interpolate(mapping, dof_handler_1, f_1, u_1);
+  VectorTools::interpolate_to_different_mesh(dof_handler_1,
+                                             u_1,
+                                             dof_handler_2,
+                                             u_2);
+
+  Point<spacedim>                 support_point, unit_support_point;
+  vector<types::global_dof_index> local_dof_indices(fe.n_dofs_per_cell());
+  deallog << "dof values of dof_handler_1:" << endl;
+  for (auto cell = dof_handler_1.begin_active(); cell != dof_handler_1.end();
```
Comment (masterleinad) [inline tests/numerics/interpolate_to_different_mesh_01.cc]: You can just use
```
for (const auto& cell: dof_handler_1.active_cell_iterators())
```

File: tests/numerics/interpolate_to_different_mesh_01.cc [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+          support_point =
+            mapping.transform_unit_to_real_cell(cell, unit_support_point);
+          deallog << " " << support_point << ":\n "
+                  << u_1[local_dof_indices[shapefun]] << endl
+                  << endl;
+        }
+    }
+  deallog << endl << "dof values of dof_handler_2:" << endl;
+  for (auto cell = dof_handler_2.begin_active(); cell != dof_handler_2.end();
+       ++cell)
```
Comment (masterleinad) [inline tests/numerics/interpolate_to_different_mesh_01.cc]:
```
for (const auto& cell: dof_handler_2.active_cell_iterators())
```

File: tests/numerics/interpolate_to_different_mesh_02.cc [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+  tria_2.refine_global(refinement_2);
+
+  hp::DoFHandler<spacedim> dof_handler_1(tria_1);
+  hp::DoFHandler<spacedim> dof_handler_2(tria_2);
+
+  hp::FECollection<spacedim> fe_collection;
+  fe_collection.push_back(FE_Q<spacedim>(2));
+  fe_collection.push_back(FE_Q<spacedim>(1));
+
+  for (auto cell = dof_handler_1.begin_active(); cell != dof_handler_1.end();
```
Comment (masterleinad) [inline tests/numerics/interpolate_to_different_mesh_02.cc]:
```
for (const auto& cell: dof_handler_1.active_cell_iterators())
```

File: tests/numerics/interpolate_to_different_mesh_02.cc [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+       ++cell)
+    cell->set_active_fe_index(0);
+
+  for (auto cell = dof_handler_2.begin_active(); cell != dof_handler_2.end();
```
Comment (masterleinad) [inline tests/numerics/interpolate_to_different_mesh_02.cc]: You can just use
```
for (const auto& cell: dof_handler_2.active_cell_iterators())
```

File: tests/numerics/interpolate_to_different_mesh_02.cc [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                                             dof_handler_2,
+                                             u_2);
+
+  Point<spacedim>                 support_point, unit_support_point;
+  vector<types::global_dof_index> local_dof_indices_1(
+    fe_collection[0].n_dofs_per_cell());
+  vector<types::global_dof_index> local_dof_indices_2(
+    fe_collection[1].n_dofs_per_cell());
+  deallog << "dof values of dof_handler_1:" << endl;
+  for (auto cell = dof_handler_1.begin_active(); cell != dof_handler_1.end();
```
Comment (masterleinad) [inline tests/numerics/interpolate_to_different_mesh_02.cc]: You can just use
```
for (const auto& cell: dof_handler_1.active_cell_iterators())
```

File: tests/numerics/interpolate_to_different_mesh_02.cc [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+          unit_support_point = cell->get_fe().unit_support_point(shapefun);
+          support_point =
+            mapping.transform_unit_to_real_cell(cell, unit_support_point);
+          deallog << " " << support_point << ":\n "
+                  << u_1[local_dof_indices_1[shapefun]] << endl
+                  << endl;
+        }
+    }
+  deallog << endl << "dof values of dof_handler_2:" << endl;
+  for (auto cell = dof_handler_2.begin_active(); cell != dof_handler_2.end();
```
Comment (masterleinad) [inline tests/numerics/interpolate_to_different_mesh_02.cc]: You can just use
```
for (const auto& cell: dof_handler_2.active_cell_iterators())
```

**Comment by starki0815** [comment]
I computed beforehand what I expected as output of the tests and checked the results with that. But your suggestion to check that interpolating from coarse to fine and back to coarse is the identity is a good additional consistency check. I'll try to add such a test.

File: tests/numerics/interpolate_to_different_mesh_01.cc (hunk 1 above)
Comment (starki0815) [reply tests/numerics/interpolate_to_different_mesh_01.cc]: Thanks, I'll change that.

**Comment by starki0815** [comment]
No worries. I have added a changelog entry now.

**Comment by masterleinad** [comment]
Thanks!

**Pull request merged**